File: SOLID/LSP/ts/VideoPlayer.ts
class VideoPlayer {
  protected volume: number = 50; // Invariant: 0 <= volume <= 100

  // Preconditions:
  // - quality must not be null

  // Postconditions:
  // - video starts playing
  play(quality: string | null): void {
    if (quality === null) {
      throw new Error("Quality cannot be null");
    }
    console.log(`Playing video in ${quality} quality`);
  }

  // Invariant preserved
  setVolume(volume: number): void {
    if (volume < 0 || volume > 100) {
      throw new Error("Invalid volume");
    }
    this.volume = volume;
  }
}

class PremiumVideoPlayer extends VideoPlayer {
  override play(quality: string | null): void {
    if (quality !== "4K") { // ❌ stronger precondition
      throw new Error("Only 4K allowed");
    }
    console.log("Playing premium 4K video");
  }
}

class BrokenVideoPlayer extends VideoPlayer {
  override play(quality: string | null): void {
    // ❌ weaker postcondition (does not play video)
    console.log("Loading video...");
  }
}

class VLCVideoPlayer extends VideoPlayer {
  boostVolume(): void {
    this.volume = 150; // ❌ breaks invariant
  }
}

File: SOLID/LSP/python/VideoPlayer.py
class VideoPlayer:
    def __init__(self):
        self._volume = 50  # Invariant: 0 <= volume <= 100

    # Preconditions:
    # - quality must not be None
    #
    # Postconditions:
    # - video starts playing
    def play(self, quality):
        if quality is None:
            raise ValueError("Quality cannot be null")
        print(f"Playing video in {quality} quality")

    # Invariant preserved
    def set_volume(self, volume):
        if volume < 0 or volume > 100:
            raise ValueError("Invalid volume")
        self._volume = volume


class PremiumVideoPlayer(VideoPlayer):
    def play(self, quality):
        if quality != "4K":  # ❌ stronger precondition
            raise ValueError("Only 4K allowed")
        print("Playing premium 4K video")


class BrokenVideoPlayer(VideoPlayer):
    def play(self, quality):
        # ❌ weaker postcondition
        print("Loading video...")


class VLCVideoPlayer(VideoPlayer):
    def boost_volume(self):
        self._volume = 150  # ❌ breaks invariant
